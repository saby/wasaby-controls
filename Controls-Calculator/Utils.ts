export default {
    prepareStringToNumber(value: string): string | null {
        const hasCarriageReturn = value.includes('\n');
        if (hasCarriageReturn) {
            return null;
        }
        const replacedValue = value.replace(/\s+/g, '');
        const parsedValue = parseFloat(replacedValue);
        return isNaN(parsedValue) ? null : parsedValue + '';
    },
};
