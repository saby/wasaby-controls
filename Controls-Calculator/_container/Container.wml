<ws:partial template="{{_options.content}}"
            attr:class="controls_Calculator_theme-{{_options.theme}}"
            scope="{{_options}}"
            on:CalculatorSetNumber="_setNumber()"
            on:CalculatorRegisterNumber="_registerNumber()"
            on:CalculatorUnregisterNumber="_unregisterNumber()"
            on:CalculatorRegisterCalculator="_registerCalculator()"
            on:CalculatorUnregisterCalculator="_unregisterCalculator()"/>
