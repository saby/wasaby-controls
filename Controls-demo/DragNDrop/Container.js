define('Controls-demo/DragNDrop/Container', [
   'UI/Base',
   'wml!Controls-demo/DragNDrop/Container/Container'
], function (Base, template) {
   'use strict';

   var ModuleClass = Base.Control.extend({
      _template: template,
      _hasList: true,
      _hasGrid: false,
      _hasTree: false,
      _hasNotes: false,
      _hasMasterDetail: false,
      _selectedKeys: []
   });
   ModuleClass._styles = ['Controls-demo/DragNDrop/Container/Container'];

   return ModuleClass;
});
