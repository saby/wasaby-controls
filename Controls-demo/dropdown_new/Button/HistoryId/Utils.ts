import { dropdownHistoryUtils } from 'Controls/dropdown';
import {
    default as HistorySourceMenu,
    getItems,
} from 'Controls-demo/dropdown_new/Button/HistoryId/historySourceMenu';

let historySources = {};

function overrideOrigSourceMethod(): void {
    dropdownHistoryUtils.getSource = (source, options) => {
        if (!historySources[options.historyId]) {
            historySources[options.historyId] = new HistorySourceMenu({});
        }
        return Promise.resolve(historySources[options.historyId]);
    };
}

function resetHistory() {
    historySources = {};
}

export { getItems, overrideOrigSourceMethod, resetHistory };
