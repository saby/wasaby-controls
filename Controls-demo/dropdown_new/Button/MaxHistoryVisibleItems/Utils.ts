import { dropdownHistoryUtils } from 'Controls/dropdown';
import {
    default as HistorySourceMenu,
    getItems,
} from 'Controls-demo/dropdown_new/Button/MaxHistoryVisibleItems/historySourceMenu';

let historySource;

function overrideOrigSourceMethod(): void {
    dropdownHistoryUtils.getSource = () => {
        if (!historySource) {
            historySource = new HistorySourceMenu({});
        }
        return Promise.resolve(historySource);
    };
}

function resetHistory() {
    historySource = undefined;
}

export { getItems, overrideOrigSourceMethod, resetHistory };
