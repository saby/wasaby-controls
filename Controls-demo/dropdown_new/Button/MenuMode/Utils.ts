import { dropdownHistoryUtils } from 'Controls/dropdown';
import { default as HistorySourceMenu } from 'Controls-demo/dropdown_new/Button/MenuMode/historySourceMenu';

let historySources = {};

function overrideOrigSourceMethod(): void {
    dropdownHistoryUtils.getSource = (source, options) => {
        if (!historySources[options.historyId]) {
            historySources[options.historyId] = new HistorySourceMenu({});
        }
        return Promise.resolve(historySources[options.historyId]);
    };
}

function resetHistory() {
    historySources = {};
}

export { overrideOrigSourceMethod, resetHistory };
