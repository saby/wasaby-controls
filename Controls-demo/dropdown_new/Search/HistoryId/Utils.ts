import { dropdownHistoryUtils } from 'Controls/dropdown';
import {
    default as HistorySourceMenu,
    getItems,
} from 'Controls-demo/dropdown_new/Search/HistoryId/historySourceMenu';

let historySource;

function overrideOrigSourceMethod(): void {
    dropdownHistoryUtils.getSource = (source) => {
        if (!historySource) {
            historySource = new HistorySourceMenu({ source });
        }
        return Promise.resolve(historySource);
    };
}

function resetHistory() {
    historySource = undefined;
}

export { getItems, overrideOrigSourceMethod, resetHistory };
