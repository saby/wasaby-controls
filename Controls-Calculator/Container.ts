import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!Controls-Calculator/_container/Container';
import Calculator from './View';
import Number from './Number';
import { SyntheticEvent } from 'UICommon/Events';

/**
 * Контрол, связывающий числа вводимые в калькулятор и сам калькулятор.
 * @extends UI/Base:Control
 * @control
 * @demo Controls-Calculator-demo/ClickOnNumber/ClickOnNumber
 * @public
 */
export default class Container extends Control {
    protected _template: TemplateFunction = template;
    private _calculators: Record<string, Calculator> = {};
    private _numbers: Record<string, Number> = {};

    private _registerCalculator(_: SyntheticEvent, instance: Calculator, id: string): void {
        this._calculators[id] = instance;

        Object.values(this._numbers).forEach((item) => {
            return item.setCalculatorRegistered(true);
        });
    }

    private _unregisterCalculator(_: SyntheticEvent, id: string): void {
        delete this._calculators[id];

        if (!Object.keys(this._calculators).length) {
            Object.values(this._numbers).forEach((item) => {
                return item.setCalculatorRegistered(false);
            });
        }
    }

    private _registerNumber(_: SyntheticEvent, instance: Number, id: string): void {
        this._numbers[id] = instance;

        if (Object.keys(this._calculators).length) {
            instance.setCalculatorRegistered(true);
        }
    }

    private _unregisterNumber(_: SyntheticEvent, id: string): void {
        delete this._numbers[id];
    }
}
