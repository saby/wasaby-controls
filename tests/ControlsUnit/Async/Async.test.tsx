/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import * as ModulesLoader from 'WasabyLoader/ModulesLoader';
import { Logger } from 'UICommon/Utils';
import { default as Async } from 'Controls/Container/Async';

describe('Controls/Container/Async', () => {
    const moduleName = 'ControlsUnit/Async/FailCallback/TestControlAsync';
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);

        jest.useFakeTimers();
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
        jest.spyOn(Logger, 'warn').mockImplementation();
        jest.spyOn(Logger, 'error').mockImplementation();
        jest.spyOn(console, 'error').mockImplementation();
        jest.spyOn(ModulesLoader, 'loadAsync').mockRejectedValue(new Error());
        jest.spyOn(ModulesLoader, 'unloadSync').mockImplementation();
    });

    afterEach(() => {
        jest.useRealTimers();
        unmountComponentAtNode(container);
        container.remove();
        jest.restoreAllMocks();
    });

    test('Асинхронная провальная загрузка на клиенте без errorCallback', async () => {
        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions={{}}
                />,
                container
            );
        });

        await act(async () => {
            jest.runAllTimers();
        });

        await act(async () => {
            jest.runAllTimers();
        });

        expect(instance.error).toMatchSnapshot();
        expect(instance.optionsForComponent.resolvedTemplate).toBeUndefined();
    });

    test('Асинхронная провальная загрузка на клиенте с errorCallback', async () => {
        const errorCallback = jest.fn();

        let instance;
        act(() => {
            render(
                <Async
                    ref={(v) => {
                        instance = v;
                    }}
                    templateName={moduleName}
                    templateOptions={{}}
                    errorCallback={errorCallback}
                />,
                container
            );
        });

        await act(async () => {
            jest.runAllTimers();
        });

        await act(async () => {
            jest.runAllTimers();
        });

        expect(instance.error).toBeUndefined();
        expect(instance.optionsForComponent.resolvedTemplate).toBeUndefined();
        expect(errorCallback).toBeCalledTimes(1);
        expect(errorCallback.mock.calls[0][0]).toBeDefined(); // первый аргумент - viewConfig
        expect(errorCallback.mock.calls[0][1]).toBeDefined(); // второй аргумент - error
    });
});
