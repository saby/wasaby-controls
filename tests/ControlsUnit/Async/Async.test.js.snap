// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls/Container/Async Асинхронная провальная загрузка на клиенте без errorCallback 1`] = `
"Ошибка загрузки контрола \\"ControlsUnit/Async/FailCallback/TestControlAsync\\"
Возможны следующие причины:
	                   • Ошибка в самом контроле
	                   • Долго отвечал БЛ метод в _beforeUpdate
	                   • Контрола не существует"
`;
